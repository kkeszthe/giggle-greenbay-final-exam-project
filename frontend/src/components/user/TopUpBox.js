import React, { useState } from 'react';
import { connect } from 'react-redux';
import { Button } from '@material-ui/core';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import PropTypes from 'prop-types';

import { updateUserBalanceAction } from '../../actions/UserActions';
import { setErrorAction } from '../../actions/ErrorActions';

function TopUpBox({ open, setOpen, topUp, setError }) {
  const [amount, setAmount] = useState(0);

  const handleClose = () => {
    setOpen(false);
  };

  const topUpClickOn = () => {
    if (!amount) {
      setError('Amount is required.');
    }
    if (!(Number.isInteger(parseInt(amount)) || parseInt(amount) < 0)) {
      setError('Amount must be whole number');
    } else {
      topUp(amount);
      handleClose();
    }
  };

  return (
    <Dialog
      open={open}
      onClose={handleClose}
      aria-labelledby="form-dialog-title"
    >
      <DialogTitle id="form-dialog-title">Top-up Account</DialogTitle>
      <DialogContent>
        <DialogContentText>
          Please provide amount you want to top-up your account with.
        </DialogContentText>
        <Grid item xs={12}>
          <TextField
            autoFocus
            margin="dense"
            id="amount"
            label="Amount"
            type="number"
            pattern="[0-9]{1,10}"
            title="Whole numbers only"
            InputProps={{ inputProps: { min: 0 } }}
            onChange={e => {
              setAmount(e.target.value);
            }}
            required
          />
        </Grid>
      </DialogContent>
      <DialogActions>
        <Button onClick={handleClose} color="primary">
          Cancel
        </Button>
        <Button type="submit" onClick={topUpClickOn} color="primary">
          Top-up
        </Button>
      </DialogActions>
    </Dialog>
  );
}

TopUpBox.propTypes = {
  open: PropTypes.bool.isRequired,
  user: PropTypes.object.isRequired,
  setOpen: PropTypes.func.isRequired,
};

const mapStateToProps = ({ user }) => {
  return { user };
};

const mapDispatchToProps = dispatch => {
  return {
    topUp: amount => {
      dispatch(updateUserBalanceAction(amount));
    },
    setError: error => {
      dispatch(setErrorAction(error));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(TopUpBox);
