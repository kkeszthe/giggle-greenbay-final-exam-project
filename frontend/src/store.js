import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';

import { userReducer } from './reducers/userReducer';
import { productsReducer } from './reducers/productsReducer';
import { tokenReducer } from './reducers/tokenReducer';
import { errorReducer } from './reducers/errorReducer';

function root(state = {}, action) {
  return {
    user: userReducer(state.user, action),
    products: productsReducer(state.troops, action),
    token: tokenReducer(state.token, action),
    error: errorReducer(state.error, action),
  };
}

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const store = createStore(
  root,
  /* preloadedState, */ composeEnhancers(applyMiddleware(thunk))
);

export { store };
