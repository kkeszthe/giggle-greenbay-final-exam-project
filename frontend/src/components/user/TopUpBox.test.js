import React from 'react';
import { Provider } from 'react-redux';
import { render, unmountComponentAtNode } from 'react-dom';
import { act, Simulate } from 'react-dom/test-utils';
import configureStore from 'redux-mock-store';

import TopUpBox from './TopUpBox';
import { store } from '../../store';

global.fetch = jest.fn(async () => {
  return await Promise.resolve({
    json: () =>
      Promise.resolve({
        products: [
          {
            id: 1,
          },
          {
            id: 2,
          },
        ],
      }),
  });
});

let container;

beforeEach(() => {
  fetch.mockClear();
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('TopUpBox tests', () => {
  it('should dispatch an action when clicked', async () => {
    const mockStore = configureStore([]);
    const mockedStore = mockStore({
      user: {
        username: 'username',
        photo_url: 'https://www.nordmet.se/wp-con…ild-2.jpg',
        balance: 100,
      },
      products: [
        {
          id: 15,
          product_name: 'practice',
          description: 'dfghjkéá-l.kg,jhvmnbdvésliretzhulkejsd',
          product_photo:
            'https://doopshop.hu/wp-content/uploads/2019/05/SPIGEN-AIRSKIN-IPHONE-XR-BLACK.jpg',
          price: 12,
          seller_id: 11,
          buyer_id: 11,
          seller_name: 'sal8',
          seller_photo:
            'https://www.nordmet.se/wp-content/uploads/2017/11/Siluett-med-spelarprofilbild-2.jpg',
        },
      ],
    });
    mockedStore.dispatch = jest.fn();
    await act(async () => {
      render(
        <Provider store={mockedStore}>
          <TopUpBox open={true} setOpen={() => {}} />
        </Provider>,
        container
      );
    });

    const button = document.getElementsByClassName('TopUpBTN')[0];
    Simulate.click(button);
    expect(mockedStore.dispatch).toHaveBeenCalledTimes(2);
  });

  it('matches snapeshot', async () => {
    Storage.prototype.getItem = jest.fn(key => {
      return 'dummy';
    });
    await act(async () => {
      render(
        <Provider store={store}>
          <TopUpBox open={true} setOpen={() => {}} />
        </Provider>,
        container
      );
    });

    expect(container.innerHTML).toMatchSnapshot();
  });
});
