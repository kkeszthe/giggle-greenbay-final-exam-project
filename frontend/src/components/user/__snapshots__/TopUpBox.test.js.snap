// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopUpBox tests matches snapeshot 1`] = `""`;
